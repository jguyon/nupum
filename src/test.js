import React, { StrictMode } from "react";
import { render as rtlRender, cleanup } from "react-testing-library";
import "jest-dom/extend-expect";
import serializer from "jest-emotion";

expect.addSnapshotSerializer(serializer);

afterEach(() => {
  cleanup();
});

export * from "react-testing-library";

export function render(ui, options) {
  const { rerender: rtlRerender, ...rest } = rtlRender(wrapUi(ui));

  function rerender(nextUi) {
    rtlRerender(wrapUi(nextUi));
  }

  return {
    ...rest,
    rerender,
  };
}

function wrapUi(ui) {
  return <StrictMode>{ui}</StrictMode>;
}
