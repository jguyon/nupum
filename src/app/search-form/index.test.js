import React from "react";
import { createMemoryHistory } from "history";
import { renderWithContext, fireEvent, act } from "../../test/react";
import SearchForm, { SearchFormProvider } from ".";

const INPUT_LABEL = "Search query";
const SUBMIT_LABEL = "Search";

test("input is initialized with empty value by default", () => {
  const { getByLabelText } = renderWithContext(
    <SearchFormProvider>
      <SearchForm />
    </SearchFormProvider>,
  );

  expect(getByLabelText(INPUT_LABEL)).toHaveAttribute("value", "");
});

test("input is initialized with correct value when on search page", () => {
  const history = createMemoryHistory({ initialEntries: ["/search?q=react"] });
  const { getByLabelText } = renderWithContext(
    <SearchFormProvider>
      <SearchForm />
    </SearchFormProvider>,
    { history },
  );

  expect(getByLabelText(INPUT_LABEL)).toHaveAttribute("value", "react");
});

test("input conserves its value when navigating", () => {
  const { history, getByLabelText, rerender, asFragment } = renderWithContext(
    <SearchFormProvider>
      <SearchForm key={1} />
    </SearchFormProvider>,
  );

  fireEvent.change(getByLabelText(INPUT_LABEL), { target: { value: "react" } });
  const fragmentBefore = asFragment();

  rerender(
    <SearchFormProvider>
      <SearchForm key={2} />
    </SearchFormProvider>,
  );

  expect(getByLabelText(INPUT_LABEL)).toHaveAttribute("value", "react");
  expect(fragmentBefore).toMatchDiffSnapshot(asFragment());

  act(() => {
    history.push("/search?q=react%20router");
  });

  expect(getByLabelText(INPUT_LABEL)).toHaveAttribute("value", "react");
  expect(fragmentBefore).toMatchDiffSnapshot(asFragment());
});

test("submitting navigates to search page", () => {
  const { history, getByLabelText } = renderWithContext(
    <SearchFormProvider>
      <SearchForm />
    </SearchFormProvider>,
  );

  fireEvent.change(getByLabelText(INPUT_LABEL), { target: { value: "react" } });
  fireEvent.click(getByLabelText(SUBMIT_LABEL));

  expect(history.location.pathname).toBe("/search");
  expect(history.location.search).toBe("?q=react");
});
