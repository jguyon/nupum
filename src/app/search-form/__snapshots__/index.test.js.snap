// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input conserves its value when navigating 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`input conserves its value when navigating 2`] = `
Snapshot Diff:
Compared values have no visual difference.
`;
